src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -2541,15 +2542,18 @@ private static float MinMaxCore<TMinMaxOperator>(ReadOnlySpan<float> x)
                 if (i != x.Length)
                 {
                     current = Vector512.LoadUnsafe(ref xRef, (uint)(x.Length - Vector512<float>.Count));
-                    if (!Vector512.EqualsAll(current, current))
+
+                    nanMask = Vector512.Equals(current, current);
+                    if (nanMask != Vector512<float>.AllBitsSet)
                     {
-                        return GetFirstNaN(current);
```

Comment:
You mentioned offline that the issue was GetFirstNaN not being inlined and that resulting in more spilling.

That spilling wasn't somehow localized to just this block but instead affected current in general?  I'd have expected that any spilling would have just been around this method, and thus would have only affected an operation that occurred once per invocation, which should have been noise at longer lengths.  That's not the case?

Reply:
It's a bit more involved than that. Vector registers have a special ABI where part of the register can be volatile vs non-volatile. This typically means that the upper bits need to be saved/restored when a call is present, even when the lower bits are not. The JIT then has specialized logic that can allow it to spill the upper bits to the lower bits of another non-volatile register.

All of this can completely change the register allocation heuristics and the spilling/caching of the rest of the method as a whole, all from one possible method call existing where data needs to be kept "live" across or into that call boundary.

In this case, the JIT ended up generating some pretty gnarly code when `GetFirstNaN` was present, with loading and caching spills along the hot path:
```asm
; This first block corresponds roughly to
;     current = Vector512.LoadUnsafe(ref xRef, (uint)i);
;     if (!Vector512.EqualsAll(current, current))
;         goto M01_L06;
;     goto M01_L07;
M01_L05:
       mov       ecx,ebp
       vmovups   zmm0,[rbx+rcx*4]
       vmovups   [rsp+80],zmm0
       vcmpeqps  k1,zmm0,[rsp+80]
       kortestw  k1,k1
       jb        short M01_L07

; This second block corresponds roughly to:
;     return GetFirstNaN(current);
M01_L06:
       lea       rcx,[rsp+80]
       call      qword ptr [7FF86AACFDE0]
       jmp       short M01_L02

; This third block corresponds to roughly
;     result = TMinMaxOperator.Invoke(result, current);
;     i += Vector512<float>.Count;
;     if (i <= oneVectorFromEnd)
;         goto M01_L05;
;     if (i == x.Length)
;         goto M01_L01;
;     goto M01_L08;
M01_L07:
       vmovups   zmm0,[rsp+0C0]
       vcmpeqps  k1,zmm0,[rsp+80]
       vpmovm2d  zmm0,k1
       vmovups   zmm1,[rsp+0C0]
       vxorps    ymm2,ymm2,ymm2
       vpcmpltd  k1,zmm1,zmm2
       vpmovm2d  zmm1,k1
       vmovups   zmm2,[rsp+80]
       vpternlogd zmm1,zmm2,[rsp+0C0],0CA
       vmovups   zmm2,[rsp+0C0]
       vmaxps    zmm2,zmm2,[rsp+80]
       vpternlogd zmm0,zmm1,zmm2,0CA
       vmovups   [rsp+0C0],zmm0
       add       ebp,10
       cmp       ebp,edi
       jle       near ptr M01_L05
       cmp       ebp,esi
       je        near ptr M01_L01
```

vs the much cleaner code without spills we get otherwise:
```asm
; This first block corresponds roughly to
;     current = Vector512.LoadUnsafe(ref xRef, (uint)i);
;     nanMask = ~Vector512.Equals(current, current);
;     if (nanMask == Vector512<float>.Zero)
;         return current.GetElement(i + IndexOfFirstMatch(nanMask));
;     goto M01_L04;
M01_L03:
       mov       r10d,r8d
       vmovups   zmm3,[rax+r10*4]
       vcmpeqps  k1,zmm3,zmm3
       vpmovm2d  zmm2,k1
       vpternlogd zmm2,zmm2,zmm2,55
       vxorps    ymm0,ymm0,ymm0
       vcmpneqps k1,zmm2,zmm0
       kortestw  k1,k1
       je        short M01_L04
       vpmovd2m  k1,zmm2
       kmovw     edx,k1
       xor       ecx,ecx
       tzcnt     rcx,rdx
       add       r8d,ecx
       cmp       r8d,10
       jae       near ptr M01_L32
       vmovups   [rsp+20],zmm3
       vmovss    xmm0,dword ptr [rsp+r8*4+20]
       jmp       near ptr M01_L01

; This second block corresponds roughly to:
;     result = TMinMaxOperator.Invoke(result, current);
;     i += Vector512<float>.Count;
;     if (i <= oneVectorFromEnd)
;         goto M01_L03;
;     if (i == x.Length)
;         goto M01_L00;
;     goto M01_L05;
M01_L04:
       vcmpeqps  k1,zmm1,zmm3
       vpmovm2d  zmm2,k1
       vxorps    ymm0,ymm0,ymm0
       vpcmpltd  k1,zmm1,zmm0
       vpmovm2d  zmm0,k1
       vpternlogd zmm0,zmm3,zmm1,0CA
       vmaxps    zmm1,zmm1,zmm3
       vpternlogd zmm2,zmm0,zmm1,0CA
       vmovaps   zmm1,zmm2
       add       r8d,10
       cmp       r8d,edx
       jle       near ptr M01_L03
       cmp       r8d,ecx
       je        near ptr M01_L00
```

Reply:
> All of this can completely change the register allocation heuristics 

Ah, got it. Thanks for the explanation.

Is there anything we can / should do to make that better in the JIT? Do we already factor such impact into the inlining heuristics?

Reply:
> Do we already factor such impact into the inlining heuristics?

I believe we have some heuristics here and there's probably more we could do in the JIT. But, this case wasn't necessarily representative because it was already duplicating a lot of work (code outside the call was repeated inside the call) and doing the analysis to match that and see we have CSE candidates making it cheaper wouldn't be trivial.

The new logic avoids duplicating such logic by breaking it down into its raw components (get a mask, see if any elements are set in the mask, then extract the index from that mask if required). This allows the helper call to be simple enough that it can be inlined and thus avoid the issue. -- `EqualsAll` basically combines the `get a mask` and `see if any elements are set in the mask` into one operation. So `GetFirstNaN` then had to repeat the `get a mask` step before it could `extract the index from that mask if required`. This made it 4-5 dense calls that make it look quite expensive.